doc/man/man5/knl.conf.5
.TH "knl.conf" "5" "Slurm Configuration File" "Slurm 25.05" "Slurm Configuration File"

.SH "NAME"
knl.conf \- Slurm configuration file for Intel Knights Landing processor.

.SH "DESCRIPTION"
This ASCII file which describes configuration information for Intel Knights
Landing processors and its name may depend upon the NodeFeatures plugin
configured in Slurm. For example, on Cray systems NodeFeatures should be configured
to "knl_cray" and its configuration file will be read from "knl_cray.conf".
The file will always be located in the same directory as the \fBslurm.conf\fR.
This file is optional.
.LP
Parameter names are case insensitive.
Any text following a "#" in the configuration file is treated
as a comment through the end of that line.
Changes to the configuration file take effect upon restart of
Slurm daemons, daemon receipt of the SIGHUP signal, or execution
of the command "scontrol reconfigure" unless otherwise noted.
.LP
The overall configuration parameters available include:

.TP
\fBAllowMCDRAM\fR
Specify the MCDRAM modes which jobs are allowed to use.
This may be a subset of MCDRAM modes supported by the node.
If not specified, all MCDRAM modes supported by the node are available for use.
The comma separated list of allowed MCDRAM modes may include any of the modes
listed below.
.IP
.RS
.TP 17
\fBcache\fR
All of MCDRAM to be used as cache.
.IP

.TP
\fBequal\fR
MCDRAM to be used partly as cache and partly combined with primary memory.
.IP

.TP
\fBflat\fR
MCDRAM to be combined with primary memory into a "flat" memory space.
.RE
.IP

.TP
\fBAllowNUMA\fR
Specify the NUMA modes which jobs are allowed to use.
This may be a subset of NUMA modes supported by the node.
If not specified, all NUMA modes supported by the node are available for use.
The comma separated list of allowed NUMA modes may include any of the modes
listed below.
Note that Slurm can only support homogeneous nodes
(e.g. the same number of cores per NUMA node).
KNL scn4 and quad modes are not homogeneous, but each NUMA mode will have
either 16 or 18 cores.
This will result in Slurm using the lower core count and finding a total of
256 threads rather than 272 threads and setting the node to a DOWN state.
Therefore it is recommended that snc4 and quad mode not be allowed at this time.
.IP
.RS
.TP 17
\fBa2a\fR
All to all
.IP

.TP
\fBsnc2\fR
Sub\-NUMA cluster 2
.IP

.TP
\fBsnc4\fR
Sub\-NUMA cluster 4
.IP

.TP
\fBhemi\fR
Hemisphere
.IP

.TP
\fBquad\fR
Quadrant
.RE
.IP

.TP
\fBAllowUserBoot\fR
A comma\-delimited list of users allowed to modify a node's MCDRAM or NUMA state.
If not specified then any user can change a node's state and reboot it.
.IP

.TP
\fBBootTime\fR
Estimated time to reboot a node in seconds.
Used as a basis for optimizing scheduling decisions.
The default value is 300 seconds (5 minutes) for the "knl_generic" plugin
and 2700 seconds (45 minutes) for the "knl_cray" plugin.
.IP

.TP
\fBCapmcPath\fR
Fully qualified path to the \fBcapmc\fR program.
The default value is "/opt/cray/capmc/default/bin/capmc".
This parameter is used only by the "knl_cray" plugin.
.IP

.TP
\fBCapmcPollFreq\fR
Time interval between when the \fBcapmc\fR program should poll for node state
changes, in seconds.
The default value is 45 seconds.
This parameter is used only by the "knl_cray" plugin.
.IP

.TP
\fBCapmcRetries\fR
Number of times to retry failed operations of the \fBcapmc\fR program.
Default value is 4.
.IP

.TP
\fBCapmcTimeout\fR
Time limit for the \fBcapmc\fR program to return status information milliseconds.
The default value is 60000 milliseconds and the minimum value is 1000 milliseconds.
This parameter is used by the "knl_cray" plugin, plus the capmc_suspend and
capmc_resume programs used for suspending and resuming nodes.
.IP

.TP
\fBCnselectPath\fR
Fully qualified path to the \fBcnselect\fR program.
The default value is "/opt/cray/sdb/default/bin/cnselect".
This parameter is used only by the "knl_cray" plugin.
.IP

.TP
\fBDefaultMCDRAM\fR
Specify the default MCDRAM modes for job's which do not specify a value.
This is only used when a node is booted and the job which has been allocated
the node does not specify a desired MCDRAM mode.
The value can include one of the possible values identified with the
\fBAllowMCDRAM\fR configuration parameter above.
The default value is "cache".
.IP

.TP
\fBDefaultNUMA\fR
Specify the default NUMA modes for job's which do not specify a value.
This is only used when a node is booted and the job which has been allocated
the node does not specify a desired NUMA mode.
The value can include one of the possible values identified with the
\fBAllowNUMA\fR configuration parameter above.
The default value is "a2a".
.IP

.TP
\fBForce\fR
If set to a non\-zero value then load the node_features/generic plugin even
on non\-KNL nodes.
Used primarily for testing purposes.
.IP

.TP
\fBLogFile\fR
Fully qualified path to a log file.
The default value is SlurmctldLogFile from the slurm.conf configuration file.
This is option is used only by the campc_suspend and campc_resume programs
(which power down and reboot nodes in the appropriate configuration).
.IP

.TP
\fBMcPath\fR
Fully qualified path to memory controller device file directory.
Children of this directory with names of the form "mc#/csrow#/ue_count"
(i.e. the count of unrecoverable memory errors) will be monitored for non\-zero
values.
If such errors are detected, the node will be set to a DOWN state and the
slurmd daemon will shutdown.
The default value is "/sys/devices/system/edac/mc".
See also \fBUmeCheckInterval\fR.
.IP

.TP
\fBNumaCpuBind\fR
Contains pairs of NUMA modes and the CpuBind mode to set a node to for that mode.
Any compute node found with or set to the specified NUMA mode will have that
node's CpuBind field set to the configured value.
The NUMA node will be followed by an equal sign the desired CpuBind mode for
that NUMA mode. Multiple NUMA mode and CpuBind modes should be in a semicolon
separated list.
By default changes to a node's NUMA mode will not effect that node's CpuBind
mode.
See the example below.
.IP

.TP
\fBSyscfgPath\fR
Fully qualified path to Intel's \fBsyscfg\fR program, which identifies current
KNL configuration by viewing BIOS settings.
If not defined, the current BIOS setting will not be available.
The default value is "/usr/bin/syscfg".
This parameter is used only by the "knl_generic" plugin.
.IP

.TP
\fBSyscfgTimeout\fR
Timeout for \fBsyscfg\fR program in milliseconds.
Default value is 1000 milliseconds.
For Dell KNL systems, experience has shown that a higher value of 10000
milliseconds is more appropriate.
.IP

.TP
\fBSystemType\fR
Used to distinguish the flavor of knl we are dealing with.
Possible options are "Dell" and "Intel".
The default value is "Intel".
This parameter is used only by the "knl_generic" plugin.
.IP

.TP
\fBUmeCheckInterval\fR
Interval, in microseconds, between checks for Uncorrectable Memory Errors (UME).
If such errors are detected, the node will be set to a DOWN state and the
slurmd daemon will shutdown.
The default value is 0 (disabled).
See also \fBMcPath\fR.
.IP

.TP
\fBValidateMode\fR
If set to 1 then validate, but do not modify the node's configured MCDRAM and
NUMA modes from the slurm.conf file. If the actual modes do not match configured
values the node will be set to a DOWN state. Every KNL nodes MCDRAM and NUMA
states must both be listed in the slurm.conf file.
This parameter is used only by the "knl_cray" plugin.
.IP

.SH "EXAMPLE"
.nf
###################################################################
# knl_cray.conf
# Slurm configuration file for Intel Knights Landing on Cray system
###################################################################
CapmcPath=/opt/cray/capmc/default/bin/capmc
CapmcTimeout=6000
DefaultMCDRAM=flat
DefaultNUMA=a2a
NumaCpuBind=a2a=core;snc2=thread;snc4=thread
LogFile=/var/tmp/slurm_node_feature.log
SyscfgPath=/usr/sbin/syscfg
.fi

.SH "COPYING"
Copyright (C) 2015\-2022 SchedMD LLC.
.LP
This file is part of Slurm, a resource management program.
For details, see <https://slurm.schedmd.com/>.
.LP
Slurm is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.
.LP
Slurm is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
details.

.SH "SEE ALSO"
.LP
\fBslurm.conf\fR(5)


NEWS
The contents of NEWS, from slurm-22.05 to the present, has been moved to
CHANGELOG.md and the CHANGELOG directory.


src/common/knl.c
/*****************************************************************************\
 *  knl.c - Infrastructure for Intel Knights Landing processor
 *****************************************************************************
 *  Copyright (C) 2015 SchedMD LLC.
 *  Written by Morris Jette <jette@schedmd.com>
 *
 *  This file is part of SLURM, a resource management program.
 *  For details, see <http://slurm.schedmd.com/>.
 *  Please also read the included file: DISCLAIMER.
 *
 *  SLURM is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  In addition, as a special exception, the copyright holders give permission
 *  to link the code of portions of this program with the OpenSSL library under
 *  certain conditions as described in each individual source file, and
 *  distribute linked combinations including the two. You must obey the GNU
 *  General Public License in all respects for all of the code used other than
 *  OpenSSL. If you modify file(s) with this exception, you may extend this
 *  exception to your version of the file(s), but you are not obligated to do
 *  so. If you do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source files in
 *  the program, then also delete it here.
 *
 *  SLURM is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with SLURM; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
\*****************************************************************************/

#if HAVE_CONFIG_H
#  include "config.h"
#  if STDC_HEADERS
#    include <string.h>
#  endif
#  if HAVE_SYS_TYPES_H
#    include <sys/types.h>
#  endif /* HAVE_SYS_TYPES_H */
#  if HAVE_UNISTD_H
#    include <unistd.h>
#  endif
#  if HAVE_INTTYPES_H
#    include <inttypes.h>
#  else /* ! HAVE_INTTYPES_H */
#    if HAVE_STDINT_H
#      include <stdint.h>
#    endif
#  endif /* HAVE_INTTYPES_H */
#else /* ! HAVE_CONFIG_H */
#  include <sys/types.h>
#  include <unistd.h>
#  include <stdint.h>
#  include <string.h>
#endif /* HAVE_CONFIG_H */

#include "slurm/slurm.h"
#include "src/common/knl.h"
#include "src/common/macros.h"
#include "src/common/parse_config.h"
#include "src/common/plugin.h"
#include "src/common/read_config.h"
#include "src/common/slurm_protocol_api.h"
#include "src/common/timers.h"
#include "src/common/xmalloc.h"
#include "src/common/xstring.h"

/*
 * WARNING:  Do not change the order of these fields or add additional
 * fields at the beginning of the structure.  If you do, KNL plugins will stop
 * working.  If you need to add fields, add them to the end of the structure.
 */
typedef struct slurm_knl_ops {
	int	(*status) (char *node_list);
	int	(*boot)	  (char *node_list, char *mcdram_type, char *numa_type);
} slurm_knl_ops_t;

/*
 * These strings must be kept in the same order as the fields
 * declared for slurm_knl_ops_t.
 */
static const char *syms[] = {
	"slurm_knl_g_status",
	"slurm_knl_g_boot"
};

static s_p_options_t knl_conf_file_options[] = {
	{"AvailNUMA", S_P_STRING},
	{"DefaultNUMA", S_P_STRING},
	{"AvailMCDRAM", S_P_STRING},
	{"DefaultMCDRAM", S_P_STRING},
	{NULL}
};

static int g_context_cnt = -1;
static slurm_knl_ops_t *ops = NULL;
static plugin_context_t **g_context = NULL;
static pthread_mutex_t g_context_lock = PTHREAD_MUTEX_INITIALIZER;
static char *knl_plugin_list = NULL;
static bool init_run = false;

static s_p_hashtbl_t *_config_make_tbl(char *filename)
{
	s_p_hashtbl_t *tbl = NULL;

	xassert(filename);

	if (!(tbl = s_p_hashtbl_create(knl_conf_file_options))) {
		error("knl.conf: %s: s_p_hashtbl_create error: %m", __func__);
		return tbl;
	}

	if (s_p_parse_file(tbl, NULL, filename, false) == SLURM_ERROR) {
		error("knl.conf: %s: s_p_parse_file error: %m", __func__);
		s_p_hashtbl_destroy(tbl);
		tbl = NULL;
	}

	return tbl;
}

/*
 * Parse knl.conf file and return available and default modes
 * avail_mcdram IN - available MCDRAM modes
 * avail_numa IN - available NUMA modes
 * default_mcdram IN - default MCDRAM mode
 * default_numa IN - default NUMA mode
 * RET - Slurm error code
 */
extern int knl_conf_read(uint16_t *avail_mcdram, uint16_t *avail_numa,
			 uint16_t *default_mcdram, uint16_t *default_numa)
{
	char *avail_mcdram_str, *avail_numa_str;
	char *default_mcdram_str, *default_numa_str;
	char *knl_conf_file, *tmp_str = NULL;
	s_p_hashtbl_t *tbl;

	/* Set default values */
	*avail_mcdram = KNL_MCDRAM_FLAG;
	*avail_numa = KNL_NUMA_FLAG;
	*default_mcdram = KNL_CACHE;
	*default_numa = KNL_ALL2ALL;

	knl_conf_file = get_extra_conf_path("knl.conf");
	if ((tbl = _config_make_tbl(knl_conf_file))) {
		if (s_p_get_string(&tmp_str, "AvailMCDRAM", tbl)) {
			*avail_mcdram = knl_mcdram_parse(tmp_str, ",");
			xfree(tmp_str);
		}
		if (s_p_get_string(&tmp_str, "AvailNUMA", tbl)) {
			*avail_numa = knl_numa_parse(tmp_str, ",");
			xfree(tmp_str);
		}
		if (s_p_get_string(&tmp_str, "DefaultMCDRAM", tbl)) {
			*default_mcdram = knl_mcdram_parse(tmp_str, ",");
			if (knl_mcdram_bits_cnt(*default_mcdram) != 1) {
				fatal("knl.conf: Invalid DefaultMCDRAM=%s",
				      tmp_str);
			}
			xfree(tmp_str);
		}
		if (s_p_get_string(&tmp_str, "DefaultNUMA", tbl)) {
			*default_numa = knl_numa_parse(tmp_str, ",");
			if (knl_numa_bits_cnt(*default_numa) != 1) {
				fatal("knl.conf: Invalid DefaultNUMA=%s",
				      tmp_str);
			}
			xfree(tmp_str);
		}
	} else {
		error("something wrong with opening/reading knl.conf");
	}
	xfree(knl_conf_file);
	s_p_hashtbl_destroy(tbl);

	avail_mcdram_str = knl_mcdram_str(*avail_mcdram);
	avail_numa_str = knl_numa_str(*avail_numa);
	default_mcdram_str = knl_mcdram_str(*default_mcdram);
	default_numa_str = knl_numa_str(*default_numa);
	if ((*default_mcdram & *avail_mcdram) == 0) {
		fatal("knl.conf: DefaultMCDRAM(%s) not within AvailMCDRAM(%s)",
		      default_mcdram_str, avail_mcdram_str);
	}
	if ((*default_numa & *avail_numa) == 0) {
		fatal("knl.conf: DefaultNUMA(%s) not within AvailNUMA(%s)",
		      default_numa_str, avail_numa_str);
	}
	debug("AvailMCDRAM=%s DefaultMCDRAM=%s",
	     avail_mcdram_str, default_mcdram_str);
	debug("AvailNUMA=%s DefaultNUMA=%s",
	     avail_numa_str, default_numa_str);
	xfree(avail_mcdram_str);
	xfree(avail_numa_str);
	xfree(default_mcdram_str);
	xfree(default_numa_str);

	return SLURM_SUCCESS;
}

/*
 * Return the count of MCDRAM bits set
 */
extern int knl_mcdram_bits_cnt(uint16_t mcdram_num)
{
	int cnt = 0, i;
	uint16_t tmp = 1;

	for (i = 0; i < 16; i++) {
		if ((mcdram_num & KNL_MCDRAM_FLAG) & tmp)
			cnt++;
		tmp = tmp << 1;
	}
	return cnt;
}

/*
 * Return the count of NUMA bits set
 */
extern int knl_numa_bits_cnt(uint16_t numa_num)
{
	int cnt = 0, i;
	uint16_t tmp = 1;

	for (i = 0; i < 16; i++) {
		if ((numa_num & KNL_NUMA_FLAG) & tmp)
			cnt++;
		tmp = tmp << 1;
	}
	return cnt;
}

/*
 * Given a KNL MCDRAM token, return its equivalent numeric value
 * token IN - String to scan
 * RET MCDRAM numeric value
 */
extern uint16_t knl_mcdram_token(char *token)
{
	uint16_t mcdram_num = 0;

	if (!strcasecmp(token, "cache"))
		mcdram_num = KNL_CACHE;
	else if (!strcasecmp(token, "flat"))
		mcdram_num = KNL_FLAT;
	else if (!strcasecmp(token, "hybrid"))
		mcdram_num = KNL_HYBRID;

	return mcdram_num;
}

/*
 * Given a KNL NUMA token, return its equivalent numeric value
 * token IN - String to scan
 * RET NUMA numeric value
 */
extern uint16_t knl_numa_token(char *token)
{
	uint16_t numa_num = 0;

	if (!strcasecmp(token, "all2all"))
		numa_num |= KNL_ALL2ALL;
	else if (!strcasecmp(token, "snc2"))
		numa_num |= KNL_SNC2;
	else if (!strcasecmp(token, "snc4"))
		numa_num |= KNL_SNC4;
	else if (!strcasecmp(token, "hemi"))
		numa_num |= KNL_HEMI;
	else if (!strcasecmp(token, "quad"))
		numa_num |= KNL_QUAD;

	return numa_num;
}

/*
 * Translate KNL MCDRAM string to equivalent numeric value
 * mcdram_str IN - String to scan
 * sep IN - token separator to search for
 * RET MCDRAM numeric value
 */
extern uint16_t knl_mcdram_parse(char *mcdram_str, char *sep)
{
	char *save_ptr = NULL, *tmp, *tok;
	uint16_t mcdram_num = 0;

	if (!mcdram_str)
		return mcdram_num;

	tmp = xstrdup(mcdram_str);
	tok = strtok_r(tmp, sep, &save_ptr);
	while (tok) {
		mcdram_num |= knl_mcdram_token(tok);
		tok = strtok_r(NULL, sep, &save_ptr);
	}
	xfree(tmp);

	return mcdram_num;
}

/*
 * Translate KNL NUMA string to equivalent numeric value
 * numa_str IN - String to scan
 * sep IN - token separator to search for
 * RET NUMA numeric value
 */
extern uint16_t knl_numa_parse(char *numa_str, char *sep)
{
	char *save_ptr = NULL, *tmp, *tok;
	uint16_t numa_num = 0;

	if (!numa_str)
		return numa_num;

	tmp = xstrdup(numa_str);
	tok = strtok_r(tmp, sep, &save_ptr);
	while (tok) {
		numa_num |= knl_numa_token(tok);
		tok = strtok_r(NULL, sep, &save_ptr);
	}
	xfree(tmp);

	return numa_num;
}

/*
 * Translate KNL MCDRAM number to equivalent string value
 * Caller must free return value
 */
extern char *knl_mcdram_str(uint16_t mcdram_num)
{
	char *mcdram_str = NULL, *sep = "";

	if (mcdram_num & KNL_CACHE) {
		xstrfmtcat(mcdram_str, "%scache", sep);
		sep = ",";
	}
	if (mcdram_num & KNL_FLAT) {
		xstrfmtcat(mcdram_str, "%sflat", sep);
		sep = ",";
	}
	if (mcdram_num & KNL_HYBRID) {
		xstrfmtcat(mcdram_str, "%shybrid", sep);
//		sep = ",";	/* Remove to avoid CLANG error */
	}

	return mcdram_str;
}

/*
 * Translate KNL NUMA number to equivalent string value
 * Caller must free return value
 */
extern char *knl_numa_str(uint16_t numa_num)
{
	char *numa_str = NULL, *sep = "";

	if (numa_num & KNL_ALL2ALL) {
		xstrfmtcat(numa_str, "%sall2all", sep);
		sep = ",";
	}
	if (numa_num & KNL_SNC2) {
		xstrfmtcat(numa_str, "%ssnc2", sep);
		sep = ",";
	}
	if (numa_num & KNL_SNC4) {
		xstrfmtcat(numa_str, "%ssnc4", sep);
		sep = ",";
	}
	if (numa_num & KNL_HEMI) {
		xstrfmtcat(numa_str, "%shemi", sep);
		sep = ",";
	}
	if (numa_num & KNL_QUAD) {
		xstrfmtcat(numa_str, "%squad", sep);
//		sep = ",";	/* Remove to avoid CLANG error */
	}

	return numa_str;

}

extern int slurm_knl_g_init(void)
{
	int rc = SLURM_SUCCESS;
	char *last = NULL, *names;
	char *plugin_type = "knl";
	char *type;

	if (init_run && (g_context_cnt >= 0))
		return rc;

	slurm_mutex_lock(&g_context_lock);
	if (g_context_cnt >= 0)
		goto fini;

	knl_plugin_list = slurm_get_knl_plugins();
	g_context_cnt = 0;
	if ((knl_plugin_list == NULL) || (knl_plugin_list[0] == '\0'))
		goto fini;

	names = knl_plugin_list;
	while ((type = strtok_r(names, ",", &last))) {
		xrealloc(ops, (sizeof(slurm_knl_ops_t) * (g_context_cnt + 1)));
		xrealloc(g_context,
			 (sizeof(plugin_context_t *) * (g_context_cnt + 1)));
		if (strncmp(type, "knl/", 4) == 0)
			type += 4; /* backward compatibility */
		type = xstrdup_printf("knl/%s", type);
		g_context[g_context_cnt] = plugin_context_create(
			plugin_type, type, (void **)&ops[g_context_cnt],
			syms, sizeof(syms));
		if (!g_context[g_context_cnt]) {
			error("cannot create %s context for %s",
			      plugin_type, type);
			rc = SLURM_ERROR;
			xfree(type);
			break;
		}

		xfree(type);
		g_context_cnt++;
		names = NULL; /* for next strtok_r() iteration */
	}
	init_run = true;

fini:
	slurm_mutex_unlock(&g_context_lock);

	if (rc != SLURM_SUCCESS)
		slurm_knl_g_fini();

	return rc;
}

extern int slurm_knl_g_fini(void)
{
	int i, j, rc = SLURM_SUCCESS;

	slurm_mutex_lock(&g_context_lock);
	if (g_context_cnt < 0)
		goto fini;

	init_run = false;
	for (i = 0; i < g_context_cnt; i++) {
		if (g_context[i]) {
			j = plugin_context_destroy(g_context[i]);
			if (j != SLURM_SUCCESS)
				rc = j;
		}
	}
	xfree(ops);
	xfree(g_context);
	xfree(knl_plugin_list);
	g_context_cnt = -1;

fini:	slurm_mutex_unlock(&g_context_lock);
	return rc;
}

extern int slurm_knl_g_status(char *node_list)
{
	DEF_TIMERS;
	int i, rc;

	START_TIMER;
	rc = slurm_knl_g_init();
	slurm_mutex_lock(&g_context_lock);
	for (i = 0; ((i < g_context_cnt) && (rc == SLURM_SUCCESS)); i++)
		rc = (*(ops[i].status))(node_list);
	slurm_mutex_unlock(&g_context_lock);
	END_TIMER2("slurm_knl_g_status");

	return rc;
}

extern int slurm_knl_g_boot(char *node_list, char *mcdram_type, char *numa_type)
{
	DEF_TIMERS;
	int i, rc;

	START_TIMER;
	rc = slurm_knl_g_init();
	slurm_mutex_lock(&g_context_lock);
	for (i = 0; ((i < g_context_cnt) && (rc == SLURM_SUCCESS)); i++)
		rc = (*(ops[i].boot))(node_list, mcdram_type, numa_type);
	slurm_mutex_unlock(&g_context_lock);
	END_TIMER2("slurm_knl_g_boot");

	return rc;
}
